sphinx_immaterial/content_tabs.py

```diff
@@ -0,0 +1,220 @@
+    }
+
+    def run(self) -> List[nodes.Node]:
+        """Run the directive."""
+        self.assert_has_content()
+        tab_set = content_tab_set(
+            "",
+            is_div=True,
+            type="md-tab-set",
+            classes=["tabbed-set", "tabbed-alternate"] + self.options.get("class", []),
+        )
+        self.set_source_info(tab_set)
+        if self.options.get("name", ""):
+            self.add_name(tab_set)
+        page_tab_set_counter[0] += 1
+        self.state.nested_parse(self.content, self.content_offset, tab_set)
+        tab_total = 0
+        for item in tab_set.children:
+            if not is_component(item, "md-tab-item"):
+                LOGGER.warning(
+                    "All children of a 'md-tab-set' should be 'md-tab-item'",
+                    location=item,
+                )
+                break
+            tab_total += 1
+        tab_set["data-tabs"] = f"{page_tab_set_counter[0]}:{tab_total}"
+        return [tab_set]
+
+
+class MaterialTabItemDirective(SphinxDirective):
+    """A single tab item in a tab set.
+    Note: This directive generates a single container,
+    for the label and content::
+        <container design_component="tab-item" has_title=True>
+            <rubric>
+                ...title nodes
+            <container design_component="tab-content">
+                ...content nodes
+    This allows for a default rendering in non-HTML outputs.
+    The ``MaterialTabSetHtmlTransform`` then transforms this container
+    into the HTML specific structure.
+    """
+
+    required_arguments = 1  # the tab label is the first argument
+    final_argument_whitespace = True
+    has_content = True
+    option_spec = {
+        "class": directives.class_option,
+    }
+
+    def run(self) -> List[nodes.Node]:
+        """Run the directive."""
+        self.assert_has_content()
+        if not is_component(self.state_machine.node, "md-tab-set"):
+            LOGGER.warning(
+                "The parent of a 'md-tab-item' should be a 'md-tab-set'",
+                location=(self.env.docname, self.lineno),
+            )
+        tab_item = nodes.container(
+            "", is_div=True, type="md-tab-item", classes=["tabbed-block"]
+        )
+
+        # add tab label
+        textnodes, _ = self.state.inline_text(self.arguments[0], self.lineno)
+        tab_label = nodes.rubric(
+            self.arguments[0],
+            *textnodes,
+            classes=["tabbed-label"],
+        )
+        self.add_name(tab_label)
+        tab_item += tab_label
+
+        # add tab content
+        tab_content = nodes.container(
+            "",
+            is_div=True,
+            type="",
+            classes=["tabbed-block"] + self.options.get("class", []),
+        )
+        self.state.nested_parse(self.content, self.content_offset, tab_content)
+        tab_item += tab_content
+
+        return [tab_item]
+
+
+class content_tab_input(nodes.Element, nodes.General):
```

Comment:
I think it would make more sense to have a `content_tab_item` node, and then handle the input and label stuff in the html translation only.

Reply:
I need the writer to explicitly use different tag names for the input & label elements. Are you saying there's a way to do that without the `stattag()` in a visitor function?

Also the attributes of a node aren't directly output as attributes of an element. These visitor functions take care of adding peculiar attributes like `for`,

Reply:
starttag just formats an HTML tag:
https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/writers/_html_base.py#l411

But why can't you use it?

Reply:
> But why can't you use it?

I think I worded my previous comment wrong. I need to use the `starttag()`. Is there a way to change the html tag without using `starttag()`?

Reply:
Basically the way the document gets converted to HTML is that the HTMLTranslator class (defined in docutils but extended by sphinx) has a `self.body` member which is just a list of strings that will get concatenated together to form the final HTML output..

Then docutils traverses the document using the walkabout function, using the HTMLTranslator object as a visitor.

In these visit methods you can just append any arbitrary text that you like, there isn't any restriction.  Additionally, in the `visit_xxx` function if you have already handled the entire content of the element you can raise `docutils.nodes.SkipNode` to prevent docutils from recursing on the children and calling the `depart_xxx` function.  So that allows you to have full control over the output.

The `starttag` method is just a utility function for formatting an HTML start tag so you can use it as you like, but you can also just format the HTML directly.  There is also a `self.attval` function for escaping an HTML attribute.

Reply:
There seems to be a lot of convenient translation (like `<` into `&lt;`) and string normalization (for class names and id attributes) that I don't want to circumvent.

Reply:
If you want to use starttag and don't have a suitable node object to pass it, you could always just create a docutils node object on the fly.

Reply:
Its hard to track what the request/comment here is concerning. Recent developments have improved the `visit_tab_set()` to factor out the `content_tab_input` class entirely. I think this should be resolved.